### PR TITLE
fix(positioning): fix popover auto position

### LINKED ISSUE #2796: fix(popover)(tooltip): "auto" position is not working correctly
**Feature**

`"auto"` positioning in the `[popover]` directive. This could be related to issue #2661, but I'm not sure.

**Bugs**

1. The popover goes off the top of the screen by about 100 pixels before it moves to the `"bottom"` position.
2. When the popover would go over the right side of the screen, it always resorts to the `"top"` position.

**Context**

version: `1.9.3`

The application is a dashboard with draggable tiles, each having a popover in the top right corner.

This is the current configuration for the popover:

- `[popover]="popoverTemplate"`
- `placement="auto"`
- `container="body"` (several parent elements have `overflow: hidden`)
- `[outsideClick]="true"`

I have created a simplified demo here: https://github.com/nfiles/ngx-bootstrap-popover-issue-demo

**Misc**

I wasn't sure where to go next in my debugging, so I started to look at the source. In `src/positioning/ng-positioning.ts` at line 197 I noticed that if the `"top"` and `"bottom"` else-if blocks were switched, everything worked as expected. I assume this is because `"top"` is preferred over `"bottom"`, but I do not know why the `"top"` position is chosen when the popover is going off screen.

### LINKED ISSUE #2998: fix(dropdown): dropdown does not take into account page boundaries
### Bug description or feature request:
If I have a button on the top right of the page which opens a dropdown, the dropdown opens but appears largely off screen. The dropdown should instead be shifted so that it is fully visible. Note, using placement option also does not correct this.

### Plunker/StackBlitz that reproduces the issue:
https://stackblitz.com/edit/ngx-bootstrap-f3spbf?file=app/app.component.html

### Versions of ngx-bootstrap, Angular, and Bootstrap:

ngx-bootstrap: 1.9.3

Angular: 4.4.6

Bootstrap: 3.3.6

### Build system: Angular CLI, System.js, webpack, starter seed:
CLI, Webpack & SystemJS

### LINKED ISSUE #3884: Sometime the tooltip is blinking annoying when open near right most screen
### Bug description or feature request:

The tooltip got blinking very annoying when open near right margin:

bug-ng-tooltip

### Versions of ngx-bootstrap, Angular, and Bootstrap:

ngx-bootstrap: 2.0.1

Angular: 5.2.1

Bootstrap:

### PR DESCRIPTION
**Pull request by YuriiKrvc**
fix popover position when 'placing="auto" and container="body"' properties are defined and the button is  placed either nearby the right side or nearby the right top corner.
fix datepicker position nearby the right side
closed #2796 
closed #3884 
closed #2998

# PR Checklist
Before creating new PR, please take a look at checklist below to make sure that you've done everything that needs to be done before we can merge it.

 - [x] read and followed the [CONTRIBUTING.md](https://github.com/valor-software/ngx-bootstrap/blob/development/CONTRIBUTING.md) guide.
 - [x] built and tested the changes locally.
 - [x] added/updated tests.
 - [ ] added/updated API documentation.
 - [ ] added/updated demos.

### CODE REVIEW

File: demo/src/app/components/+datepicker/demos/basic/basic.html [hunk 1]
```diff
@@ -5,11 +5,13 @@
            class="form-control"
            bsDatepicker>
   </div>
-  <div class="col-xs-12 col-12 col-md-4 form-group">
+  <div class="pull-right form-group">
```
Comment (przemyslawjanpietrzak) [inline demo/src/app/components/+datepicker/demos/basic/basic.html:8]: first child of row has col class and second not. Both should have it.

File: src/positioning/ng-positioning.ts [hunk 2]
```diff
@@ -98,6 +98,13 @@ export class Positioning {
     let placementPrimary = placement.split(' ')[0] || 'top';
     const placementSecondary = placement.split(' ')[1] || 'center';
 
+    const rightOffset = window.innerWidth
+                        - targetElement.getBoundingClientRect().left
```
Comment (przemyslawjanpietrzak) [inline src/positioning/ng-positioning.ts:102]: move targetElement.getBoundingClientRect() to variable

File: src/positioning/ng-positioning.ts [hunk 3]
```diff
@@ -149,14 +158,43 @@ export class Positioning {
             parseFloat(targetElStyles.marginBottom));
         targetElPosition.bottom +=
           hostElPosition.top - targetElPosition.height;
+
+        if(rightOffset < 100
```
Comment (przemyslawjanpietrzak) [inline src/positioning/ng-positioning.ts:162]: pls move 100, 200 and 1.6 to constants and name it

**Comment by Domainv** [comment]
We need complete refactor of position service.
Point fixes are insufficient

**Comment by odedkeren1** [comment]
We will be happy to have this fix as well !!!
Looking forward

**Pull request closed without merging**